bin/generator: Big rewrite

Insti opened the pull request:
Can now generate tests without updating the test version number.

New command line options:
```
Usage: bin/generate [options] exercise-generator
	-f, --freeze                     Don't update test version
	-h, --help                       Displays this help message
	-v, --verbose                    Display progress messages
```

`sha1` in the "Common test data version: <%= sha1 %>" comment is now the
commit ID of the relevant canonical-data.json file

Files will only be written if they have changed.

Split code up into classes.
100% test coverage.

Added dependency: `require_all` gem
    
Just requiring files in (potentially random) directory order was not sufficient to satisfy all class dependencies. The `require_all` gem ensures dependencies are met as the files are `require`d

bmulvihill commented:
@Insti this looks cool, I'll see if I can take a look today!

bmulvihill commented on Gemfile:
```diff
@@ -4,3 +4,5 @@ gem 'minitest'
 gem 'rubocop', '0.36.0'
 gem 'simplecov'
 gem 'rake'
+gem 'values'
```
I am always hesitant to add new dependencies to a project, do these gems provide anything that is essential and we can't handle ourselves?

Insti replied:
Good point.

It depends on where you want to draw the line, the 2 I added were convenient for adding the functionality I wanted. I'll look into what it would take to eliminate the dependencies. I suspect not much.

bmulvihill replied:
Yeah I don't suspect these 2 would cause any major issues - but I just know from experience dependency creep can be a big problem!

bmulvihill commented on lib/generator/files.rb:
```diff
@@ -0,0 +1,44 @@
+      end
+    end
+
+    class MutableFile < ImmutableFile
```
Does inheritance here make sense?  Does this pass the is a? test for inheritance, a MutableFile is a ImmutableFile?

Insti replied:
This may be a case of misleading naming.

It's basically some files need to have a save method, and others don't. 
This way the intention of which files are meant to be `save`able is more clear.

It may not actually be important, and might just be a case of "just don't call save on the ones you don't want to save". 

Do you have any suggestions on how to handle this?

Doing the inheritance the other way around seems wrong if we need to define a save method which we then disable.

There's surely an elegant way of handling this hiding in there somewhere.

bmulvihill replied:
Would `class WritableFile < ReadableFile` make more sense?

bmulvihill commented on lib/generator/files/exercise_files.rb:
```diff
@@ -0,0 +1,83 @@
+
+    end
+
+    # Assumes methods exist: paths, exercise_name
```
Maybe we can add those methods into this module which raise NotImplementedError to ensure classes which include this module are in compliance

```ruby
  def paths
    raise NotImplementError
  end
```

Insti replied:
Excellent point. Thanks.

This sort of thing is what I love about code reviews. :heart:

Insti commented:
Thanks @bmulvihill, Was that all you found or did you need to stop part way through?

bmulvihill commented:
@Insti I got pulled away midway through, I will try and finish looking it over this evening!

bmulvihill commented on lib/generator/files/exercise_files.rb:
```diff
@@ -0,0 +1,83 @@
+module Generator
+  module Files
+    module GeneratorCases
+      def self.available(track_path)
```
I would consider using `extend self` here since this seems to be more of a utility module, that would eliminate the need for self references
```ruby
module GeneratorCases
  extend self
```

Insti replied:
The style guide seems to suggest `module_function` rather than `extend_self` so I'll use `module_function` instead: https://github.com/bbatsov/ruby-style-guide#module-function

bmulvihill commented:
Made some more small comments, overall I think this looks good and I like that there is test coverage now + more options when generating a test! 👍

Insti commented:
- Addressed feedback provided by @bmulvihill.
- Applied some minor renamingss which will hopefully make things clearer and more consistent.

Anything (other than squashing) that needs to be done before we merge?

@kotp, @tommyschaefer : Have you had a chance to look over this?

tommyschaefer commented on lib/generator/template_values.rb:
```diff
@@ -0,0 +1,48 @@
+  end
+
+  # This is not a standalone module, is that a problem?
+  module TemplateValuesFactory
```
Are the `TODO`s in this module meant to be completed after merging?

I think it's completely reasonable if so, especially since these changes make the process of generating test sooooooo much more pleasant (`--freeze` 😍).

Just wanted to double check that these didn't get missed 😄

Insti replied:
Yes, the TODOs are all things that are intended to be separate follow-up PRs as they involve modifications to the exercise files.

kotp commented:
@insti let's squash what you want together and make this easier to do a "last review" on it?

Insti commented:
Ready for final review. Let's get this merged. Any objections?
@kotp, @tommyschaefer, @bmulvihill

tommyschaefer commented on bin/generate:
```diff
@@ -3,34 +3,9 @@
-klass = Object.const_get(cases.split('_').map(&:capitalize).join)
-Generator.new(exercise, klass).generate
+generator = Generator::CommandLine.new(paths).parse(ARGV)
+exit 1 unless generator
```
Would it be reasonable to add a little message here if generation is successful? Something simple maybe like the message you get when you initialize a new git repository.

kotp replied:
It may be that silence signals success.

Insti replied:
Default is totally silent.
But if you use -v you get a success message.

tommyschaefer replied:
Ah! Thats awesome! I totally missed that. Thank you 😄

Insti replied:
I'd like to change that, so the default was to print progress messages and a --quiet flag would make it silent, but that's for a future PR. I was trying to keep the functionality of this version as close to the one it was replacing as possible. (But some other things crept in.)

kotp replied:
I really like the "silence is golden" approach.  Noisy if you ask for it.

kotp commented on lib/generator/command_line.rb:
```diff
@@ -0,0 +1,119 @@
+    end
+
+    def validate_options
+      return true unless @options[:help]
```
Since this is a method that is required to return a boolean, the name could be `valid_options?`  This communicates that usage.

kotp commented on lib/generator/command_line.rb:
```diff
@@ -0,0 +1,119 @@
+    def validate_options
+      return true unless @options[:help]
+      $stdout.puts usage
+      false
```
puts returns nil, and so would be sufficient to let the line above be the return.

kotp commented on lib/generator/command_line.rb:
```diff
@@ -0,0 +1,119 @@
+      return true if @options[:exercise_name]
+      $stderr.puts "Exercise name required!\n"
+      $stdout.puts usage
+      false
```
Same here for query name and puts returning nil.

kotp commented on lib/generator/command_line.rb:
```diff
@@ -0,0 +1,119 @@
+    def validate_cases
+      return true if available_generators.include?(@options[:exercise_name])
+      $stderr.puts "A generator does not currently exist for #{@options[:exercise_name]}!"
+      false
```
Seeing a pattern here.

kotp commented on lib/generator/command_line.rb:
```diff
@@ -0,0 +1,119 @@
+    end
+
+    def usage
+      option_parser.help +
```
Prefer using `<<` to concatenate the strings, as it uses a single string, appropriately.  One string for the garbage collector.

kotp commented on lib/generator/files.rb:
```diff
@@ -0,0 +1,56 @@
+    end
+
+    class Writable < Readable
+      def save_if_changed(new_content)
```
The name should be `save`.  We don't care if it is changed or not, but for efficiency, internally, it can make that distinction.  Also, if it is new content, it is not changed.  `content` should be sufficient.  Unless it is truly requiring that it is indeed new content.

Insti replied:
good points, fixed.

kotp commented on lib/generator/files.rb:
```diff
@@ -0,0 +1,56 @@
+        File.open(filename, 'w') do |file|
+          file.write new_content
+        end
+      end
```
same comments as save_if_changed.

kotp commented on lib/generator/template_values.rb:
```diff
@@ -0,0 +1,48 @@
+    end
+  end
+
+  # This is not a standalone module, is that a problem?
```
Comment is likely ready to be removed.  Or at least the question part of it, if it is for documenting/guidance.

kotp commented on lib/generator/template_values.rb:
```diff
@@ -0,0 +1,48 @@
+      compensated_version = version - 1
+
+      TemplateValues.new(
+        # TODO: rename sha1 to abbreviated_commit_hash
```
This TODO appearse to be done.

kotp commented on test/fixtures/xruby/exercises/alpha/example.tt:
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env ruby
```
This is a fixture, the folder as an example could be named `foo` or `bar` as traditionally used as examples, even working examples.

`alpha` may suggest there will be a `beta` and `delta`, etc.

Insti replied:
I thought there might have been when I created it, but it turned out further letters weren't needed.

kotp commented on test/fixtures/xruby/lib/alpha_cases.rb:
```diff
@@ -0,0 +1,20 @@
+class AlphaCase < OpenStruct
```
`FooCase` or `BarCase` due to the heavy stroop effect on the word "alpha".

Insti replied:
Although that is also the name of a computer hardware manufacturer..
Naming is hard.

kotp replied:
This is a sanity test, really, right?  Could be `SanityCase`.

kotp commented on test/generator/files/track_files_test.rb:
```diff
@@ -0,0 +1,117 @@
+
+    class TestsVersionFileTest < Minitest::Test
+      class TestTestsVersionFile < TestsVersionFile
+        def save_if_changed(new_content)
```
This will hopefully just be save, as mentioned in prior comment.

kotp commented on test/generator/template_values_test.rb:
```diff
@@ -0,0 +1,59 @@
+  class TemplateValuesFactoryTest < Minitest::Test
+    class TestTemplateValuesFactory
+      def exercise_name
+        'alpha'
```
Consider `foo` or `bar` name instead.

kotp commented on lib/generator/command_line.rb:
```diff
@@ -95,8 +95,8 @@ def option_parser
     def usage
-      option_parser.help +
-        "\nAvailable exercise generators:\n" +
+      option_parser.help <<
```
Quick safety check on the value that is returned from the `help` method.  The `<<` that I was referencing was the second concatenation, rather than both.  The reason is I did not peek at the help method, to make sure that modifying that in place would not cause a change that affects other calls to help.

Insti replied:
`option_parser.help` returns a brand new string on every call, so no problems there.

Insti commented:
:tada: Hooray!

Thanks everybody for your reviews/feedback and help! :heart:
